[DirectX] Implement Shader Flag Analysis for `UAVsAtEveryStage`

Icohedron opened the pull request:
Fixes #112272

In addition to the implementation of the UAVsAtEveryStage shader flag analysis, several unrelated tests have had the `dx.valver` module metadata defined to avoid setting the UAVsAtEveryStage shader flag in them.
Example:
```
!dx.valver = !{!0}
!0 = !{i32 1, i32 8}
```

Icohedron committed:
Implement UAVsAtEveryStage shader flag

Icohedron committed:
Change vertex shader test function name and remove hlsl.export from it

bogner commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -265,6 +265,26 @@ void ModuleShaderFlags::initialize(Module &M, DXILResourceTypeMap &DRTM,
+          MMDI.ShaderProfile != Triple::EnvironmentType::Compute &&
+          MMDI.ShaderProfile != Triple::EnvironmentType::Pixel;
+    else // MMDI.ValidatorVersion >= VersionTuple(1, 8)
+      switch (MMDI.ShaderProfile) {
```
I think it'd be clearer if we just put all of the shader stages in the switch, rather than having an extra "if" based on the shader version. Also, we should really be using more functions here - `ModuleShaderFlags::initialize` is getting very long.

A helper like so would make this dead simple:
```c++
bool hasUAVsAtEveryStage(DXILResourceMap &DRM, const ModuleMetadataInfo &MMDI) {
  if (DRM.uavs().empty())
    return false;

  switch (MMDI.ShaderProfile) {
  default:
    return false;
  case Triple::EnvironmentType::Compute:
  case Triple::EnvironmentType::Pixel:
    return false;
  case Triple::EnvironmentType::Vertex:
  case Triple::EnvironmentType::Geometry:
  case Triple::EnvironmentType::Hull:
  case Triple::EnvironmentType::Domain:
    return true;
  case Triple::EnvironmentType::Library:
  case Triple::EnvironmentType::RayGeneration:
  case Triple::EnvironmentType::Intersection:
  case Triple::EnvironmentType::AnyHit:
  case Triple::EnvironmentType::ClosestHit:
  case Triple::EnvironmentType::Miss:
  case Triple::EnvironmentType::Callable:
  case Triple::EnvironmentType::Mesh:
  case Triple::EnvironmentType::Amplification:
    return MMDI.ValidatorVersion < VersionTuple(1, 8);
  }
}
```

bogner commented on llvm/test/CodeGen/DirectX/ShaderFlags/max-64-uavs-array-sm6_5.ll:
```diff
@@ -29,5 +29,10 @@ define void @test() "hlsl.export" {
   ret void
 }
 
+; Set dx.valver and dx.resmayalias to prevent flags ResMayNotAlias and
+; UAVsAtEveryStage from being set, as to not distract from the flag that is
+; actually being tested
```
I don't think we need this comment in every test file. Just mention this in the PR description / commit message and people can find it in the git history if they're trying to figure out why valver was added here.

bogner commented on llvm/test/CodeGen/DirectX/ShaderFlags/uavs-at-every-stage-lib-valver1.7.ll:
```diff
@@ -0,0 +1,39 @@
+; DXC: - Name:            SFI0
+; DXC-NEXT:     Size:            8
+; DXC-NEXT:     Flags:
+; DXC-NOT:   {{[A-Za-z]+: +true}}
```
I don't think there's much value in explicitly checking that other flags are *not* set here - that's just likely to cause code churn as we add more flags isn't it?

Icohedron replied:
Yea, I think I agree.
A separate PR should also be made to remove similar checks from other shader flag tests like [doubles.ll](https://github.com/llvm/llvm-project/blob/3e4e365fac0503d9fcd9bb34ee1c27bf17d95e3a/llvm/test/CodeGen/DirectX/ShaderFlags/doubles.ll#L25-L27)

Icohedron committed:
Simplify the logic for setting UAVsAtEveryStage

- Move the logic for setting UAVsAtEveryStage out into its own function to not increase the length of ModuleShaderFlags::initialize
- Use one switch statement to handle both special cases with regards to DXIL validator version

Co-authored-by: Justin Bogner <mail@justinbogner.com>

Icohedron committed:
Remove repeated comments among tests

Icohedron committed:
Remove checks for no other shader flag is being set

Icohedron committed:
Merge remote-tracking branch 'upstream/main' into shader-flag-UAVsAtEveryStage

Icohedron committed:
Add additional fixes to unrelated shader-flag tests to account for UAVsAtEveryStage

Icohedron committed:
Apply clang-format

inbelic commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -32,6 +32,35 @@
+  case Triple::EnvironmentType::Geometry:
+  case Triple::EnvironmentType::Hull:
+  case Triple::EnvironmentType::Domain:
+    return true;
```
Shouldnt this be only if the version is greater than 1.8, based on the issue description?

Icohedron replied:
If the version is less than 1.8, they would still get the UAVsAtEveryStage flag because they are not Compute or Pixel

Icohedron replied:
I think the real potential issue is if in the future we get new shader types, the UAVsAtEveryStage flag will not get assigned to the new shader type when the DXIL version is < 1.8

Icohedron replied:
@bogner Do you think the `hasUAVsAtEveryStage` function should be this instead?
```c++
bool hasUAVsAtEveryStage(DXILResourceMap &DRM, const ModuleMetadataInfo &MMDI) {
  if (DRM.uavs().empty())
    return false;

  switch (MMDI.ShaderProfile) {
  default:
    return MMDI.ValidatorVersion < VersionTuple(1, 8);
  case Triple::EnvironmentType::Compute:
  case Triple::EnvironmentType::Pixel:
    return false;
  case Triple::EnvironmentType::Vertex:
  case Triple::EnvironmentType::Geometry:
  case Triple::EnvironmentType::Hull:
  case Triple::EnvironmentType::Domain:
    return true;
  }
}
```
This is so that we don't have to edit the switch every time a new shader stage gets added.

bogner replied:
This is one of the places where the fact that we're using the triple environment rather than a dedicated enum is kind of annoying. If we had a dedicated enum then we would get a `-Wcovered-switch` warning if that came up and my answer to this question would unambiguously be no.

However, I still don't think that this would be an improvement. Consider:
- If we get another shader stage then by definition the validator version is greater than 1.8, since a new shader stage couldn't be added without bumping the DXIL version.
- Whether or not UAVsAtEveryStage should be set for the new shader stage probably depends on what that stage is, so we'd probably want to modify this anyway.
- If we have a triple that doesn't have a shader stage as its environment (ie, it isn't a `dxil-` triple at all), then calling this function doesn't make any sense. We could arguably make the `default` case `llvm_unreachable()` and just crash if that were to happen.
